# SQL CARTO QUERY FUNCTION
# written for PUI2016_Python2
#### and make it python 2 and 3 compatible by Hao Xi.

##### This notebook is set up to link as a default to the fb55 account.  To turn in the homework use the same account you used in the lab (hvt201)  and query the database that you were querying in class (citibike_feb_2015, named as citibike)

In [6]:
SQL_SOURCE = 'https://hx517.carto.com/api/v2/sql?q='

try:
    from urllib.request import urlopen
except ImportError:
    from urllib2 import urlopen
try:
    from urllib.error import HTTPError
except ImportError:
    from urllib2 import HTTPError
try:
    from urllib.parse import urlencode
except ImportError:
    from urllib import urlencode
import io
import ast
import pandas as pd
# try to make it python 2 and 3 compatible.

def queryCartoDB(query, format='CSV', source=SQL_SOURCE):
    '''queries carto datasets from a given carto account
    Arguments: 
    query - string: a valid sql query string
    format - outlut format  OPTIONAL (default CSV)
    source - a valid sql api endpoint OPTIONAL (default carto fb55 account)
    Returns:
    the return of the sql query AS A STRING
    NOTES:
    designed for the carto API, tested only with CSV return format'''
    
    data = urllib.urlencode({'format': format, 'q': query})
    try:
        response = urllib2.urlopen(source, data)
    except urllib2.HTTPError, e:
        raise ValueError('\n'.join(ast.literal_eval(e.readline())['error']))
    except Exception:
        raise
    return response.read()

In [7]:
test = '''
SELECT *
FROM citibike
'''

In [8]:
pd.read_csv(StringIO.StringIO(queryCartoDB(test)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  tripduration  \
0            5960       NaN                   NaN           351   
1            9806       NaN                   NaN           643   
2           13113       NaN                   NaN          1379   
3               1       NaN                   NaN           801   
4               2       NaN                   NaN           379   
5               3       NaN                   NaN          2474   
6               4       NaN                   NaN           818   
7               5       NaN                   NaN           544   
8               6       NaN                   NaN           717   
9               7       NaN                   NaN          1306   
10              8       NaN                   NaN           913   
11              9       NaN                   NaN           759   
12             10       NaN                   NaN           585   
13             11       NaN                   NaN           581   
14             12       NaN                   NaN           204   
15             13       NaN                   NaN          1169   
16             14       NaN                   NaN           419   
17             15       NaN                   NaN           527   
18             16       NaN                   NaN          1013   
19             17       NaN                   NaN           587   
20             18       NaN                   NaN           525   
21             19       NaN                   NaN           670   
22             20       NaN                   NaN           711   
23             21       NaN                   NaN          1559   
24             22       NaN                   NaN           683   
25             23       NaN                   NaN           522   
26             24       NaN                   NaN           389   
27             25       NaN                   NaN           276   
28             26       NaN                   NaN           907   
29             27       NaN                   NaN           946   
...           ...       ...                   ...           ...   
46170       46171       NaN                   NaN           812   
46171       46172       NaN                   NaN           647   
46172       46173       NaN                   NaN           563   
46173       46174       NaN                   NaN           470   
46174       46175       NaN                   NaN           299   
46175       46176       NaN                   NaN           629   
46176       46177       NaN                   NaN           499   
46177       46178       NaN                   NaN           474   
46178       46179       NaN                   NaN           345   
46179       46180       NaN                   NaN           829   
46180       46181       NaN                   NaN           776   
46181       46182       NaN                   NaN           697   
46182       46183       NaN                   NaN          1703   
46183       46184       NaN                   NaN           922   
46184       46185       NaN                   NaN           849   
46185       46186       NaN                   NaN           826   
46186       46187       NaN                   NaN           473   
46187       46188       NaN                   NaN           332   
46188       46189       NaN                   NaN           456   
46189       46190       NaN                   NaN           327   
46190       46191       NaN                   NaN           306   
46191       46192       NaN                   NaN          1248   
46192       46193       NaN                   NaN          1165   
46193       46194       NaN                   NaN           993   
46194       46195       NaN                   NaN          1422   
46195       46196       NaN                   NaN           689   
46196       46197       NaN                   NaN           428   
46197       46198       NaN                   NaN     

## Task 1¶
### a. Sort data by start_station_id - only checking trips with duration <= 3 hours
#### 1. the top 10 records

In [11]:
test = '''
SELECT * FROM citibike
WHERE tripduration<10800
ORDER BY start_station_id
LIMIT 10
'''

In [12]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  tripduration  \
0        3302       NaN                   NaN           569   
1        1527       NaN                   NaN           962   
2        2142       NaN                   NaN           339   
3        2282       NaN                   NaN           301   
4         387       NaN                   NaN           961   
5         649       NaN                   NaN           344   
6        1230       NaN                   NaN           934   
7        1983       NaN                   NaN          1944   
8        5136       NaN                   NaN           455   
9        3957       NaN                   NaN           387   

                starttime                stoptime  start_station_id  \
0  2015-02-01 14:52:00+00  2015-02-01 15:01:00+00                72   
1  2015-02-01 11:47:00+00  2015-02-01 12:03:00+00                72   
2  2015-02-01 12:54:00+00  2015-02-01 13:00:00+00                72   
3  2015-02-01 13:08:00+00  2015-02-01 13:13:00+00                72   
4  2015-02-01 08:38:00+00  2015-02-01 08:54:00+00                72   
5  2015-02-01 09:49:00+00  2015-02-01 09:54:00+00                72   
6  2015-02-01 11:14:00+00  2015-02-01 11:30:00+00                72   
7  2015-02-01 12:39:00+00  2015-02-01 13:12:00+00                72   
8  2015-02-01 17:55:00+00  2015-02-01 18:03:00+00                72   
9  2015-02-01 15:57:00+00  2015-02-01 16:03:00+00                72   

  start_station_name  start_station_latitude  start_station_longitude  \
0   W 52 St & 11 Ave               40.767272               -73.993929   
1   W 52 St & 11 Ave               40.767272               -73.993929   
2   W 52 St & 11 Ave               40.767272               -73.993929   
3   W 52 St & 11 Ave               40.767272               -73.993929   
4   W 52 St & 11 Ave               40.767272               -73.993929   
5   W 52 St & 11 Ave               40.767272               -73.993929   
6   W 52 St & 11 Ave               40.767272               -73.993929   
7   W 52 St & 11 Ave               40.767272               -73.993929   
8   W 52 St & 11 Ave               40.767272               -73.993929   
9   W 52 St & 11 Ave               40.767272               -73.993929   

   end_station_id         end_station_name  end_station_latitude  \
0             499       Broadway & W 60 St             40.769155   
1             212  W 16 St & The High Line             40.743349   
2             478         11 Ave & W 41 St             40.760301   
3             447          8 Ave & W 52 St             40.763707   
4             212  W 16 St & The High Line             40.743349   
5            2021          W 45 St & 8 Ave             40.759291   
6             212  W 16 St & The High Line             40.743349   
7             328  Watts St & Greenwich St             40.724055   
8             529          W 42 St & 8 Ave             40.757570   
9             514         12 Ave & W 40 St             40.760875   

   end_station_longitude  bikeid    usertype  birth_year  gender  
0             -73.981918   17429  Subscriber      1981.0       1  
1             -74.006818   16524  Subscriber      1977.0       1  
2             -73.998842   19672  Subscriber      1982.0       1  
3             -73.985162   18174  Subscriber      1974.0       1  
4             -74.006818   20786  Subscriber      1972.0       2  
5             -73.988597   14692  Subscriber      1986.0       2  
6             -74.006818   17541  Subscriber      1969.0       1  
7             -74.009660   17849    Customer         NaN       0  
8             -73.990985   16427  Subscriber      1983.0       1  
9             -74.002777   15282  Subscriber      1991.0       1

#### 2. the last 10 records

In [13]:
test = '''
SELECT * FROM citibike
WHERE tripduration<10800
ORDER BY start_station_id DESC
LIMIT 10
'''

In [14]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  tripduration  \
0        1228       NaN                   NaN          1900   
1        1863       NaN                   NaN           689   
2        1472       NaN                   NaN           683   
3        1506       NaN                   NaN          7225   
4        1587       NaN                   NaN           858   
5        1628       NaN                   NaN           745   
6         293       NaN                   NaN           301   
7        1333       NaN                   NaN          1243   
8        2205       NaN                   NaN           790   
9        2026       NaN                   NaN           307   

                starttime                stoptime  start_station_id  \
0  2015-02-01 11:14:00+00  2015-02-01 11:45:00+00              3002   
1  2015-02-01 12:25:00+00  2015-02-01 12:36:00+00              3002   
2  2015-02-01 11:43:00+00  2015-02-01 11:55:00+00              3002   
3  2015-02-01 11:46:00+00  2015-02-01 13:47:00+00              3002   
4  2015-02-01 11:53:00+00  2015-02-01 12:07:00+00              3002   
5  2015-02-01 11:57:00+00  2015-02-01 12:09:00+00              3002   
6  2015-02-01 07:49:00+00  2015-02-01 07:54:00+00              3002   
7  2015-02-01 11:27:00+00  2015-02-01 11:48:00+00              3002   
8  2015-02-01 13:01:00+00  2015-02-01 13:14:00+00              3002   
9  2015-02-01 12:42:00+00  2015-02-01 12:47:00+00              3002   

           start_station_name  start_station_latitude  \
0  South End Ave & Liberty St               40.711512   
1  South End Ave & Liberty St               40.711512   
2  South End Ave & Liberty St               40.711512   
3  South End Ave & Liberty St               40.711512   
4  South End Ave & Liberty St               40.711512   
5  South End Ave & Liberty St               40.711512   
6  South End Ave & Liberty St               40.711512   
7  South End Ave & Liberty St               40.711512   
8  South End Ave & Liberty St               40.711512   
9  South End Ave & Liberty St               40.711512   

   start_station_longitude  end_station_id               end_station_name  \
0               -74.015756             525               W 34 St & 11 Ave   
1               -74.015756             358  Christopher St & Greenwich St   
2               -74.015756             308        St James Pl & Oliver St   
3               -74.015756             499             Broadway & W 60 St   
4               -74.015756             355          Bayard St & Baxter St   
5               -74.015756             369          Washington Pl & 6 Ave   
6               -74.015756             417         Barclay St & Church St   
7               -74.015756             360           William St & Pine St   
8               -74.015756             368             Carmine St & 6 Ave   
9               -74.015756             147       Greenwich St & Warren St   

   end_station_latitude  end_station_longitude  bikeid    usertype  \
0             40.755942             -74.002116   17849    Customer   
1             40.732916             -74.007114   19309  Subscriber   
2             40.713079             -73.998512   16987  Subscriber   
3             40.769155             -73.981918   21202    Customer   
4             40.716021             -73.999744   15748  Subscriber   
5             40.732241             -74.000264   20901  Subscriber   
6             40.712912             -74.010202   19484  Subscriber   
7             40.707179             -74.008873   18288    Customer   
8             40.730386             -74.002150   17004  Subscriber   
9             40.715422             -74.011220   18386  Subscriber   

   birth_year  gender  
0         NaN       0  
1      1996.0       1  
2      1980.0       2  
3         NaN       0  
4      1959.0       1  
5      1980.0       1  
6      1980.0       2  
7         NaN       0  
8      1979.0       1  
9      1984.0       2

### b. Sort data by tripduration - only checking trips with duration <= 3 hours
#### 1. the top 10 records

In [15]:
test = '''
SELECT * FROM citibike
WHERE tripduration<10800
ORDER BY tripduration
LIMIT 10
'''

In [16]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  tripduration  \
0       13043       NaN                   NaN            60   
1       14117       NaN                   NaN            60   
2        8284       NaN                   NaN            60   
3       11104       NaN                   NaN            60   
4        3110       NaN                   NaN            60   
5       13482       NaN                   NaN            60   
6        7299       NaN                   NaN            60   
7        7798       NaN                   NaN            60   
8       21407       NaN                   NaN            60   
9       19243       NaN                   NaN            60   

                starttime                stoptime  start_station_id  \
0  2015-02-04 06:54:00+00  2015-02-04 06:55:00+00               521   
1  2015-02-04 08:41:00+00  2015-02-04 08:42:00+00               296   
2  2015-02-03 08:07:00+00  2015-02-03 08:08:00+00               251   
3  2015-02-03 18:01:00+00  2015-02-03 18:02:00+00               471   
4  2015-02-01 14:34:00+00  2015-02-01 14:35:00+00               317   
5  2015-02-04 07:52:00+00  2015-02-04 07:53:00+00               512   
6  2015-02-02 17:14:00+00  2015-02-02 17:15:00+00               474   
7  2015-02-02 21:39:00+00  2015-02-02 21:40:00+00               379   
8  2015-02-04 22:38:00+00  2015-02-04 22:39:00+00               523   
9  2015-02-04 18:22:00+00  2015-02-04 18:23:00+00               248   

        start_station_name  start_station_latitude  start_station_longitude  \
0          8 Ave & W 31 St               40.750450               -73.994811   
1     Division St & Bowery               40.714131               -73.997047   
2      Mott St & Prince St               40.723180               -73.994800   
3  Grand St & Havemeyer St               40.712868               -73.956981   
4        E 6 St & Avenue B               40.724537               -73.981854   
5          W 29 St & 9 Ave               40.750073               -73.998393   
6          5 Ave & E 29 St               40.745168               -73.986831   
7          W 31 St & 7 Ave               40.749156               -73.991600   
8          W 38 St & 8 Ave               40.754666               -73.991382   
9    Laight St & Hudson St               40.721854               -74.007718   

   end_station_id         end_station_name  end_station_latitude  \
0             521          8 Ave & W 31 St             40.750450   
1             296     Division St & Bowery             40.714131   
2             251      Mott St & Prince St             40.723180   
3             471  Grand St & Havemeyer St             40.712868   
4             317        E 6 St & Avenue B             40.724537   
5             512          W 29 St & 9 Ave             40.750073   
6             474          5 Ave & E 29 St             40.745168   
7             379          W 31 St & 7 Ave             40.749156   
8             523          W 38 St & 8 Ave             40.754666   
9             248    Laight St & Hudson St             40.721854   

   end_station_longitude  bikeid    usertype  birth_year  gender  
0             -73.994811   14788  Subscriber        1971       1  
1             -73.997047   21644  Subscriber        1990       1  
2             -73.994800   19896  Subscriber        1982       1  
3             -73.956981   16340  Subscriber        1980       1  
4             -73.981854   20572  Subscriber        1985       1  
5             -73.998393   21627  Subscriber        1960       2  
6             -73.986831   20842  Subscriber        1956       1  
7             -73.991600   17643  Subscriber        1986       1  
8             -73.991382   16148  Subscriber        1980       1  
9             -74.007718   20309  Subscriber        1970       1

#### 2. the last 10 records

In [19]:
test = '''
SELECT * FROM citibike
WHERE tripduration<10800
ORDER BY tripduration DESC
LIMIT 10
'''

In [20]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  tripduration  \
0       41392       NaN                   NaN         10757   
1       10912       NaN                   NaN         10660   
2       40221       NaN                   NaN         10174   
3        7472       NaN                   NaN          9969   
4        9696       NaN                   NaN          9938   
5       15336       NaN                   NaN          9807   
6       25203       NaN                   NaN          9711   
7       19897       NaN                   NaN          9572   
8        7073       NaN                   NaN          9556   
9       11810       NaN                   NaN          9373   

                starttime                stoptime  start_station_id  \
0  2015-02-07 12:17:00+00  2015-02-07 15:16:00+00               284   
1  2015-02-03 17:42:00+00  2015-02-03 20:40:00+00               512   
2  2015-02-07 09:31:00+00  2015-02-07 12:20:00+00               405   
3  2015-02-02 18:22:00+00  2015-02-02 21:08:00+00               386   
4  2015-02-03 14:10:00+00  2015-02-03 16:55:00+00               468   
5  2015-02-04 10:53:00+00  2015-02-04 13:36:00+00               435   
6  2015-02-05 11:12:00+00  2015-02-05 13:54:00+00               457   
7  2015-02-04 19:09:00+00  2015-02-04 21:48:00+00               402   
8  2015-02-02 13:25:00+00  2015-02-02 16:05:00+00               499   
9  2015-02-03 19:25:00+00  2015-02-03 22:01:00+00               324   

              start_station_name  start_station_latitude  \
0          Greenwich Ave & 8 Ave               40.739017   
1                W 29 St & 9 Ave               40.750073   
2  Washington St & Gansevoort St               40.739323   
3           Centre St & Worth St               40.714948   
4             Broadway & W 55 St               40.765265   
5                W 21 St & 6 Ave               40.741740   
6             Broadway & W 58 St               40.766953   
7             Broadway & E 22 St               40.740343   
8             Broadway & W 60 St               40.769155   
9        DeKalb Ave & Hudson Ave               40.689888   

   start_station_longitude  end_station_id               end_station_name  \
0               -74.002638             173             Broadway & W 49 St   
1               -73.998393             435                W 21 St & 6 Ave   
2               -74.008119             405  Washington St & Gansevoort St   
3               -74.002345             264           Maiden Ln & Pearl St   
4               -73.981923            2023        E 55 St & Lexington Ave   
5               -73.994156             435                W 21 St & 6 Ave   
6               -73.981693            2021                W 45 St & 8 Ave   
7               -73.989551             517          Pershing Square South   
8               -73.981918             499             Broadway & W 60 St   
9               -73.981013             324        DeKalb Ave & Hudson Ave   

   end_station_latitude  end_station_longitude  bikeid    usertype  \
0             40.760647             -73.984427   15082  Subscriber   
1             40.741740             -73.994156   16099  Subscriber   
2             40.739323             -74.008119   21323  Subscriber   
3             40.707065             -74.007319   19921  Subscriber   
4             40.759681             -73.970314   19070  Subscriber   
5             40.741740             -73.994156   18088  Subscriber   
6             40.759291             -73.988597   14726  Subscriber   
7             40.751581             -73.977910   14691  Subscriber   
8             40.769155             -73.981918   14802  Subscriber   
9             40.689888             -73.981013   15018  Subscriber   

   birth_year  gender  
0        1973       1  
1        1950       1  
2        1983       1  
3        1987       2  
4        1970       1  
5        1985       1  
6        1962       1  
7        1966       2  
8        1940       1  
9      

### c. List all unique start_station_id values

In [21]:
test = '''
SELECT DISTINCT start_station_id FROM citibike
ORDER BY start_station_id
'''

In [22]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

start_station_id  Unnamed: 1
0                  72         NaN
1                  79         NaN
2                  82         NaN
3                  83         NaN
4                 116         NaN
5                 119         NaN
6                 120         NaN
7                 127         NaN
8                 128         NaN
9                 137         NaN
10                143         NaN
11                144         NaN
12                146         NaN
13                147         NaN
14                150         NaN
15                151         NaN
16                152         NaN
17                153         NaN
18                157         NaN
19                160         NaN
20                161         NaN
21                164         NaN
22                167         NaN
23                168         NaN
24                173         NaN
25                174         NaN
26                195         NaN
27                212         NaN
28                216         NaN
29                217         NaN
..                ...         ...
297               525         NaN
298               526         NaN
299               527         NaN
300               528         NaN
301               529         NaN
302               530         NaN
303               531         NaN
304               532         NaN
305               533         NaN
306               534         NaN
307               537         NaN
308               538         NaN
309               539         NaN
310               545         NaN
311               546         NaN
312              2000         NaN
313              2001         NaN
314              2002         NaN
315              2004         NaN
316              2005         NaN
317              2006         NaN
318              2008         NaN
319              2009         NaN
320              2010         NaN
321              2012         NaN
322              2017         NaN
323              2021         NaN
324              2022         NaN
325              2023         NaN
326              3002         NaN

[327 rows x 2 columns]

### d. Count the number/average/min/max of trips

In [24]:
test = '''
SELECT COUNT(tripduration) as numtrips,
       AVG(tripduration) as avgtrips,
       MIN(tripduration) as mintrips,
       MAX(tripduration) as maxtrips
FROM citibike
'''

In [25]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

numtrips    avgtrips  mintrips  maxtrips
0     46200  675.865823        60     43016

## Task 2
### a. Selecting trips started on Feb-02-2015 only

In [27]:
test = '''
SELECT * FROM citibike
WHERE starttime >= '2015-02-02 00:00'
AND   starttime  < '2015-02-03 00:00'
'''

In [28]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  tripduration  \
0           6442       NaN                   NaN           199   
1           6443       NaN                   NaN           418   
2           6444       NaN                   NaN           276   
3           6445       NaN                   NaN           420   
4           6446       NaN                   NaN           304   
5           6447       NaN                   NaN           622   
6           6448       NaN                   NaN           371   
7           6449       NaN                   NaN           477   
8           6450       NaN                   NaN           743   
9           6451       NaN                   NaN           599   
10          6452       NaN                   NaN           345   
11          6453       NaN                   NaN           721   
12          6454       NaN                   NaN           182   
13          6455       NaN                   NaN           154   
14          6456       NaN                   NaN           435   
15          6457       NaN                   NaN           310   
16          6458       NaN                   NaN          1274   
17          6459       NaN                   NaN           642   
18          6460       NaN                   NaN           256   
19          6461       NaN                   NaN           334   
20          6462       NaN                   NaN            90   
21          6463       NaN                   NaN           203   
22          6464       NaN                   NaN           649   
23          6465       NaN                   NaN           319   
24          6466       NaN                   NaN           154   
25          6467       NaN                   NaN           550   
26          6468       NaN                   NaN           268   
27          6469       NaN                   NaN           153   
28          6470       NaN                   NaN           309   
29          6471       NaN                   NaN           138   
...          ...       ...                   ...           ...   
1429        7871       NaN                   NaN           722   
1430        7872       NaN                   NaN           274   
1431        7873       NaN                   NaN           951   
1432        7874       NaN                   NaN           276   
1433        7875       NaN                   NaN           426   
1434        7876       NaN                   NaN           578   
1435        7877       NaN                   NaN          2621   
1436        7878       NaN                   NaN           621   
1437        7879       NaN                   NaN           323   
1438        7880       NaN                   NaN           393   
1439        7881       NaN                   NaN           597   
1440        7882       NaN                   NaN           244   
1441        7883       NaN                   NaN          1023   
1442        7884       NaN                   NaN           419   
1443        7885       NaN                   NaN          1429   
1444        7886       NaN                   NaN           676   
1445        7887       NaN                   NaN           879   
1446        7888       NaN                   NaN           593   
1447        7889       NaN                   NaN           839   
1448        7890       NaN                   NaN           560   
1449        7891       NaN                   NaN           165   
1450        7892       NaN                   NaN           813   
1451        7893       NaN                   NaN           764   
1452        7894       NaN                   NaN          2111   
1453        7895       NaN                   NaN           518   
1454        7896       NaN                   NaN           371   
1455        7897       NaN                   NaN           188   
1456        7898       NaN                   NaN           689   
1457        7899       NaN                   N

### b. Selecting trips started on the weekends and calculate the average trip duration during weekends

In [29]:
test = '''
SELECT * FROM citibike
WHERE extract(DOW FROM starttime) in (0,6)
'''

In [30]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  tripduration  \
0            5960       NaN                   NaN           351   
1               1       NaN                   NaN           801   
2               2       NaN                   NaN           379   
3               3       NaN                   NaN          2474   
4               4       NaN                   NaN           818   
5               5       NaN                   NaN           544   
6               6       NaN                   NaN           717   
7               7       NaN                   NaN          1306   
8               8       NaN                   NaN           913   
9               9       NaN                   NaN           759   
10             10       NaN                   NaN           585   
11             11       NaN                   NaN           581   
12             12       NaN                   NaN           204   
13             13       NaN                   NaN          1169   
14             14       NaN                   NaN           419   
15             15       NaN                   NaN           527   
16             16       NaN                   NaN          1013   
17             17       NaN                   NaN           587   
18             18       NaN                   NaN           525   
19             19       NaN                   NaN           670   
20             20       NaN                   NaN           711   
21             21       NaN                   NaN          1559   
22             22       NaN                   NaN           683   
23             23       NaN                   NaN           522   
24             24       NaN                   NaN           389   
25             25       NaN                   NaN           276   
26             26       NaN                   NaN           907   
27             27       NaN                   NaN           946   
28             28       NaN                   NaN           144   
29             29       NaN                   NaN           257   
...           ...       ...                   ...           ...   
13201       46171       NaN                   NaN           812   
13202       46172       NaN                   NaN           647   
13203       46173       NaN                   NaN           563   
13204       46174       NaN                   NaN           470   
13205       46175       NaN                   NaN           299   
13206       46176       NaN                   NaN           629   
13207       46177       NaN                   NaN           499   
13208       46178       NaN                   NaN           474   
13209       46179       NaN                   NaN           345   
13210       46180       NaN                   NaN           829   
13211       46181       NaN                   NaN           776   
13212       46182       NaN                   NaN           697   
13213       46183       NaN                   NaN          1703   
13214       46184       NaN                   NaN           922   
13215       46185       NaN                   NaN           849   
13216       46186       NaN                   NaN           826   
13217       46187       NaN                   NaN           473   
13218       46188       NaN                   NaN           332   
13219       46189       NaN                   NaN           456   
13220       46190       NaN                   NaN           327   
13221       46191       NaN                   NaN           306   
13222       46192       NaN                   NaN          1248   
13223       46193       NaN                   NaN          1165   
13224       46194       NaN                   NaN           993   
13225       46195       NaN                   NaN          1422   
13226       46196       NaN                   NaN           689   
13227       46197       NaN                   NaN           428   
13228       46198       NaN                   NaN     

In [31]:
test = '''
SELECT AVG(tripduration) as avgtrips FROM citibike
WHERE extract(DOW FROM starttime) in (0,6)
'''

In [32]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

avgtrips  Unnamed: 1
0  662.942181         NaN

So the average trip duration time during weekends is 662.94 s

### c. do the same for weekday

In [35]:
test = '''
SELECT * FROM citibike
WHERE extract(DOW FROM starttime) not in (0,6)
'''

In [36]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

cartodb_id  the_geom  the_geom_webmercator  tripduration  \
0            9806       NaN                   NaN           643   
1           13113       NaN                   NaN          1379   
2            6442       NaN                   NaN           199   
3            6443       NaN                   NaN           418   
4            6444       NaN                   NaN           276   
5            6445       NaN                   NaN           420   
6            6446       NaN                   NaN           304   
7            6447       NaN                   NaN           622   
8            6448       NaN                   NaN           371   
9            6449       NaN                   NaN           477   
10           6450       NaN                   NaN           743   
11           6451       NaN                   NaN           599   
12           6452       NaN                   NaN           345   
13           6453       NaN                   NaN           721   
14           6454       NaN                   NaN           182   
15           6455       NaN                   NaN           154   
16           6456       NaN                   NaN           435   
17           6457       NaN                   NaN           310   
18           6458       NaN                   NaN          1274   
19           6459       NaN                   NaN           642   
20           6460       NaN                   NaN           256   
21           6461       NaN                   NaN           334   
22           6462       NaN                   NaN            90   
23           6463       NaN                   NaN           203   
24           6464       NaN                   NaN           649   
25           6465       NaN                   NaN           319   
26           6466       NaN                   NaN           154   
27           6467       NaN                   NaN           550   
28           6468       NaN                   NaN           268   
29           6469       NaN                   NaN           153   
...           ...       ...                   ...           ...   
32939       39381       NaN                   NaN           321   
32940       39382       NaN                   NaN           449   
32941       39383       NaN                   NaN           120   
32942       39384       NaN                   NaN           977   
32943       39385       NaN                   NaN           501   
32944       39386       NaN                   NaN           441   
32945       39387       NaN                   NaN           388   
32946       39388       NaN                   NaN           607   
32947       39389       NaN                   NaN           284   
32948       39390       NaN                   NaN           271   
32949       39391       NaN                   NaN           572   
32950       39392       NaN                   NaN           498   
32951       39393       NaN                   NaN           425   
32952       39394       NaN                   NaN           267   
32953       39395       NaN                   NaN           165   
32954       39396       NaN                   NaN           751   
32955       39397       NaN                   NaN           747   
32956       39398       NaN                   NaN           511   
32957       39399       NaN                   NaN           489   
32958       39400       NaN                   NaN          1183   
32959       39401       NaN                   NaN           649   
32960       39402       NaN                   NaN           377   
32961       39403       NaN                   NaN           485   
32962       39404       NaN                   NaN           298   
32963       39405       NaN                   NaN           648   
32964       39406       NaN                   NaN          1261   
32965       39407       NaN                   NaN           331   
32966       39408       NaN                   NaN     

In [37]:
test = '''
SELECT AVG(tripduration) as avgtrips FROM citibike
WHERE extract(DOW FROM starttime) not in (0,6)
'''

In [38]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

avgtrips  Unnamed: 1
0  681.052292         NaN

The average trip duration for weekday is 662.94 s

## Task 3
### a. Showing the list of start station locations

In [44]:
test = '''
SELECT CDB_TransformToWebmercator(CDB_LatLng(start_station_latitude, start_station_longitude)) 
             as the_geom_webmercator,
       MIN(cartodb_id) as cartodb_id,
       MIN(start_station_name) as start_station_name 
FROM citibike
GROUP BY start_station_latitude, start_station_longitude
'''

In [45]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

the_geom_webmercator  cartodb_id  \
0    0101000020110F0000863F88A4096A5FC19D1DAFD5DFF1...          84   
1    0101000020110F0000BB2EC9F82D6A5FC19150E1B461FB...         412   
2    0101000020110F0000D62721D3CA6D5FC157A5D529F6F5...         516   
3    0101000020110F000060197671C96C5FC197A854AF46FA...         359   
4    0101000020110F0000993A68067B6C5FC1B222DA492BFA...          78   
5    0101000020110F00004FF9F3E90D6D5FC1F284479FC8F9...        1019   
6    0101000020110F00003816CC930D6B5FC17B393D28A0F9...         110   
7    0101000020110F00006B6C9CBF186A5FC109C57CA0D8FC...           7   
8    0101000020110F00008B8D725EB16B5FC1BF41718A0DF6...         460   
9    0101000020110F0000BA97F1D5D36C5FC13D326BC3E4F5...         799   
10   0101000020110F000052716DFB7B6D5FC1634AEE724BF4...        5444   
11   0101000020110F000079B8DA32AB685FC194DBF126F9F4...         459   
12   0101000020110F000004115CC3E56B5FC1D8A47A41AFF7...         618   
13   0101000020110F00006C38A18AC36A5FC1CCDA8D8DAFFC...         288   
14   0101000020110F0000EC774642E56C5FC13DF29213A3F7...          26   
15   0101000020110F00006EAC10AF716A5FC1491D46B773FC...         519   
16   0101000020110F0000F211F9C1C7695FC1E298B1F0A7F1...         707   
17   0101000020110F00002FE7B4B3FC6D5FC150F152E3E5F5...        1205   
18   0101000020110F0000C12D0F658D675FC1323CC0ADCAF2...           6   
19   0101000020110F00007392D1EEBD685FC17A448F6F1AF2...        8955   
20   0101000020110F00003CBD186C2C6A5FC1C2D62BADBCF6...         370   
21   0101000020110F0000B8ECD5128D6D5FC1CF0946E36EF6...         797   
22   0101000020110F00007BC267B2266C5FC14EDB28B171F2...         153   
23   0101000020110F0000C74600C9BE6B5FC1451EF62F18F9...         424   
24   0101000020110F0000F9471F1F9A6C5FC125B4B2DACDF8...          55   
25   0101000020110F0000B4B6D9A8FF695FC1B42BC0556EF1...        3237   
26   0101000020110F00005293059FE96B5FC1283F96AE66FD...         387   
27   0101000020110F0000F9420761D76C5FC19CAEC2F187F6...        1151   
28   0101000020110F00001EF7B62F966B5FC1649B856FF3FA...         252   
29   0101000020110F00002011DE6425685FC115AB984DDAF1...         653   
..                                                 ...         ...   
297  0101000020110F00003A962D08646B5FC1C2CAB91C36F2...         574   
298  0101000020110F0000BECCC59977685FC1A94F3336D6F2...         968   
299  0101000020110F00009492832C826A5FC13DFED2D04CF2...         115   
300  0101000020110F0000CECBC1AD456B5FC149916BC5DDFD...        1386   
301  0101000020110F00002E55C204246D5FC1CA6FCB5F1FFB...         306   
302  0101000020110F00005A738F401D6A5FC1E98304C839F9...         218   
303  0101000020110F00000DCC625D696D5FC158A6C354E2F6...        2184   
304  0101000020110F0000E9C5F392EA6A5FC14849234B6AF3...         805   
305  0101000020110F000029CC2D4A18685FC1860D5E9389F1...         410   
306  0101000020110F0000B056ADFF5C6A5FC1500051A8DCF1...         533   
307  0101000020110F0000416B223CE96B5FC1EE8BEFEA70FA...         411   
308  0101000020110F000049FA5A0D3A695FC1BA4AA87F7AF1...        2402   
309  0101000020110F000055982AD3F16B5FC1607FD69C04F6...         339   
310  0101000020110F0000A452F7AC7B6B5FC185F5C472A5F6...          83   
311  0101000020110F000001D7D2EC996C5FC1E4C65FCA5FF8...         216   
312  0101000020110F00003410F605196D5FC18E052E1DD2F8...        1092   
313  0101000020110F00006F4EDEDB7D6A5FC177CDA056BEF6...          81   
314  0101000020110F00001288289AB06A5FC143D788E4D9FC...         303   
315  0101000020110F00000AFD8FC8C36C5FC15AF2587A04FB...         176   
316  0101000020110F00001607D538556B5FC119A6CA6F41FC...         124   
317  0101000020110F0000049C1A69CE6C5FC1205139A81BF8...          13   
318  0101000020110F00002FD56A5DDF685FC1A458A08E56F1...         765   
319  0101000020110F00002D1B19CE146C5FC11E9608E2CDF3...        1737   
320  0101000020110F0000E463F723266A5FC14F49CEE530FB...        2481   
321  0101000020110F00005E9509B1416B5FC19122492B05FD...          71   
322  0101000020110F000020CE41FA366A5FC1E47EC

### b. Showing the numbers of trips started per station but only for stations within 500m of Time Square

In [46]:
test = '''
SELECT CDB_TransformToWebmercator(
                CDB_LatLng(start_station_latitude, 
                           start_station_longitude)
       ) as the_geom_webmercator, 
       MIN(cartodb_id) as cartodb_id,
       COUNT(tripduration) as numtrips,
       MIN(start_station_name) as start_station_name
FROM citibike
WHERE ST_DWithin(
        CDB_LatLng(start_station_latitude,
                   start_station_longitude)::geography,
        CDB_LatLng(40.7577, -73.9857)::geography,
      500)
GROUP BY start_station_latitude, start_station_longitude
'''

In [47]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

the_geom_webmercator  cartodb_id  numtrips  \
0  0101000020110F0000F22A632FBE6A5FC1363A28CFADFB...        1115       112   
1  0101000020110F000004AF971C1D6B5FC17A48F3C1A8FB...          54       251   
2  0101000020110F00001607D538556B5FC119A6CA6F41FC...         124       141   
3  0101000020110F00003F9300B2976B5FC1B8B88F3102FC...         783       221   
4  0101000020110F0000F55695027D6B5FC186E5BB69D7FB...          33       507   
5  0101000020110F000020D6742CE16A5FC112F4D73D73FC...          42       213   
6  0101000020110F00009BB87D02B76A5FC146C144E9E5FB...          19       141   

   start_station_name  
0     W 43 St & 6 Ave  
1  Broadway & W 41 St  
2     W 45 St & 8 Ave  
3     W 42 St & 8 Ave  
4     W 41 St & 8 Ave  
5  Broadway & W 49 St  
6     W 45 St & 6 Ave

## Task 4   
## Find the station that had the longest average trip duration during weekends and within 500m of TimeSquare!

In [48]:
test = '''
SELECT CDB_TransformToWebmercator(
                CDB_LatLng(start_station_latitude, 
                           start_station_longitude)
       ) as the_geom_webmercator, 
       MIN(cartodb_id) as cartodb_id,
       COUNT(tripduration) as numtrips,
       AVG(tripduration) as avgtrips,
       MIN(start_station_name) as start_station_name
FROM citibike
WHERE ST_DWithin(
        CDB_LatLng(start_station_latitude,
                   start_station_longitude)::geography,
        CDB_LatLng(40.7577, -73.9857)::geography,
      500) and extract(DOW FROM starttime) in (0,6)
GROUP BY start_station_latitude, start_station_longitude
ORDER BY avgtrips DESC
'''

In [49]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

the_geom_webmercator  cartodb_id  numtrips  \
0  0101000020110F000020D6742CE16A5FC112F4D73D73FC...          42        48   
1  0101000020110F00001607D538556B5FC119A6CA6F41FC...         124        44   
2  0101000020110F000004AF971C1D6B5FC17A48F3C1A8FB...          54        33   
3  0101000020110F00009BB87D02B76A5FC146C144E9E5FB...          19        15   
4  0101000020110F0000F55695027D6B5FC186E5BB69D7FB...          33        73   
5  0101000020110F0000F22A632FBE6A5FC1363A28CFADFB...        1115        10   
6  0101000020110F00003F9300B2976B5FC1B8B88F3102FC...         783        28   

      avgtrips  start_station_name  
0  1010.104167  Broadway & W 49 St  
1   762.931818     W 45 St & 8 Ave  
2   683.121212  Broadway & W 41 St  
3   675.400000     W 45 St & 6 Ave  
4   643.260274     W 41 St & 8 Ave  
5   629.700000     W 43 St & 6 Ave  
6   579.142857     W 42 St & 8 Ave

So the station that had the longest average trip duration during weekends and within 500m of TimeSquare is Broadway & W 49 St.

## Extra for SQL
##     Create lines for trips started from stations within 500m of Times Squares and lasted less than 2 hours. 
## The number of trips per each pair of stations are output as attributes of these lines.

In [50]:
test = '''
SELECT ST_MakeLine(CDB_TransformToWebmercator(CDB_LatLng(start_station_latitude,start_station_longitude)),
             CDB_TransformToWebmercator(CDB_LatLng(end_station_latitude,end_station_longitude)))
                   as the_geom_webmercator,
     MIN(cartodb_id) as cartodb_id,
     COUNT(tripduration) as numtrips,
     MIN(start_station_name) as start_station_name,
     MIN(end_station_name) as end_station_name
FROM citibike
WHERE ST_DWithin(
      CDB_LatLng(start_station_latitude,
            start_station_longitude)::geography,
      CDB_LatLng(40.7577,-73.9857)::geography,
    500)
    AND tripduration<7200
GROUP BY start_station_latitude, start_station_longitude, end_station_latitude, end_station_longitude
'''

In [51]:
pd.read_csv(io.BytesIO(queryCartoDB(test)), sep=',')

the_geom_webmercator  cartodb_id  numtrips  \
0    0102000020110F000002000000F22A632FBE6A5FC1363A...       12043         1   
1    0102000020110F0000020000009BB87D02B76A5FC146C1...       34407         1   
2    0102000020110F00000200000004AF971C1D6B5FC17A48...       30188         1   
3    0102000020110F0000020000003F9300B2976B5FC1B8B8...        7566         5   
4    0102000020110F00000200000020D6742CE16A5FC112F4...        8850         4   
5    0102000020110F0000020000003F9300B2976B5FC1B8B8...       31004         1   
6    0102000020110F00000200000020D6742CE16A5FC112F4...        4636         2   
7    0102000020110F000002000000F55695027D6B5FC186E5...        6788         6   
8    0102000020110F000002000000F22A632FBE6A5FC1363A...       20181         2   
9    0102000020110F0000020000001607D538556B5FC119A6...       26631         1   
10   0102000020110F000002000000F22A632FBE6A5FC1363A...       36993         1   
11   0102000020110F000002000000F55695027D6B5FC186E5...       11242         1   
12   0102000020110F0000020000003F9300B2976B5FC1B8B8...       22083         3   
13   0102000020110F0000020000003F9300B2976B5FC1B8B8...       13700         2   
14   0102000020110F000002000000F55695027D6B5FC186E5...        8798         1   
15   0102000020110F0000020000003F9300B2976B5FC1B8B8...       14842         1   
16   0102000020110F00000200000004AF971C1D6B5FC17A48...       38207         2   
17   0102000020110F0000020000009BB87D02B76A5FC146C1...       14181         2   
18   0102000020110F0000020000001607D538556B5FC119A6...       19813         1   
19   0102000020110F000002000000F55695027D6B5FC186E5...       36988         1   
20   0102000020110F0000020000001607D538556B5FC119A6...       43818         1   
21   0102000020110F000002000000F55695027D6B5FC186E5...         876         3   
22   0102000020110F000002000000F55695027D6B5FC186E5...       15276         2   
23   0102000020110F0000020000009BB87D02B76A5FC146C1...       18472         2   
24   0102000020110F000002000000F22A632FBE6A5FC1363A...        1115         2   
25   0102000020110F0000020000009BB87D02B76A5FC146C1...       14749         3   
26   0102000020110F000002000000F55695027D6B5FC186E5...       26955         2   
27   0102000020110F00000200000020D6742CE16A5FC112F4...       21504         1   
28   0102000020110F00000200000004AF971C1D6B5FC17A48...       21258         2   
29   0102000020110F00000200000020D6742CE16A5FC112F4...       45898         1   
..                                                 ...         ...       ...   
542  0102000020110F0000020000001607D538556B5FC119A6...       15912         3   
543  0102000020110F00000200000020D6742CE16A5FC112F4...        9169         7   
544  0102000020110F000002000000F55695027D6B5FC186E5...        6319        13   
545  0102000020110F00000200000020D6742CE16A5FC112F4...         570         7   
546  0102000020110F0000020000009BB87D02B76A5FC146C1...       13614         3   
547  0102000020110F0000020000009BB87D02B76A5FC146C1...       20724         1   
548  0102000020110F000002000000F55695027D6B5FC186E5...       12873         3   
549  0102000020110F000002000000F55695027D6B5FC186E5...       13148         5   
550  0102000020110F00000200000020D6742CE16A5FC112F4...       29300         2   
551  0102000020110F00000200000004AF971C1D6B5FC17A48...       19451         5   
552  0102000020110F000002000000F22A632FBE6A5FC1363A...       37508         1   
553  0102000020110F0000020000003F9300B2976B5FC1B8B8...       12273         1   
554  0102000020110F00000200000004AF971C1D6B5FC17A48...       11481         2   
555  0102000020110F00000200000004AF971C1D6B5FC17A48...       11191         1   
556  0102000020110F000002000000F22A632FBE6A5FC1363A...       11880         2   
557  0102000020110F00000200000004AF971C1D6B5FC17A48...        1080         7   
558  0102000020110F00000200000004AF971C1D6B5FC17A48...       12908         2   
559  0102000020110F000002000000F55695027D6B5FC186E5...       31209         1   
560  0102000020110F000002000000F55

# extra credit: make the function python 2 and 3 compatible so that it works on the  PUI2016_Python3 kernel

##### I've made the function python 2 and 3 compatible.